### ISSUE TITLE
Add support for temperature, top-k, and top-p sampling in text generation

### ISSUE THREAD

**Issue by ishikawa** [ishikawa/q#42]
Currently, the text generation implementation uses greedy sampling, which always selects the token with the highest probability. While simple and deterministic, this often leads to repetitive and less diverse outputs.

To improve generation quality and flexibility, I suggest adding support for:

- **Temperature sampling**: Scales logits before softmax. Higher values increase diversity, while lower values make output more deterministic. If `temperature=0`, we should fall back to greedy sampling to avoid division-by-zero errors.
- **Top-k sampling**: Randomly samples from the top-k most probable tokens.
- **Top-p (nucleus) sampling**: Samples from the smallest set of tokens whose cumulative probability exceeds a threshold `p`.

### Suggested Implementation (pseudo-code):

```python
def sample_token(logits, temperature=1.0, top_k=None, top_p=None):
    if temperature == 0.0:
        # Fall back to greedy sampling
        return int(mx.argmax(logits))

    # Apply temperature
    logits = logits / temperature
    probs = mx.softmax(logits)

    # Top-k filtering
    if top_k is not None:
        topk_indices = mx.topk(logits, k=top_k, sorted=False)
        probs = probs[topk_indices]
        probs /= probs.sum()
        return int(mx.random.choice(topk_indices, p=probs))

    # Top-p filtering
    if top_p is not None:
        sorted_indices = mx.argsort(logits)[::-1]
        sorted_probs = probs[sorted_indices]
        cumulative_probs = mx.cumsum(sorted_probs)
        cutoff = mx.where(cumulative_probs > top_p)[0][0] + 1
        top_p_indices = sorted_indices[:cutoff]
        top_p_probs = sorted_probs[:cutoff]
        top_p_probs /= top_p_probs.sum()
        return int(mx.random.choice(top_p_indices, p=top_p_probs))

    # Pure temperature sampling
    return int(mx.random.choice(mx.arange(len(probs)), p=probs))
```

It would be great to allow users to pass these parameters (`temperature`, `top_k`, `top_p`) when generating tokens.
